### PR TITLE
Fix links and implemenet html_writer

### PR DESCRIPTION
**Pull request by owenherbert**
* Fixed broken moodle_urls
* Implemented html_writer

### CODE REVIEW

File: journaltodiaryxfr.php [hunk 1]
```diff
@@ -232,22 +232,18 @@
                      name="button1"
                      onClick="return clClick()"
                      type="submit" value="'
-                     .get_string('transfer', 'diary').'"> <a href="'.$url2->out(false).'"
-
-                     class="btn btn-secondary"
-                     style="border-radius: 8px">'
-                     .get_string('cancel', 'diary').'</a></input>';
+                     .get_string('transfer', 'diary').'"> '
+                     . html_writer::link($url2, get_string('transfer', 'diary'), ['class' => 'btn btn-secondary', 'role' => 'button', 'style' => 'border-radius: 8px']) .'</input>';
```
Comment (matthewhilton) [inline journaltodiaryxfr.php]: Should the string used in the button be `cancel` and not `transfer` ?

File: journaltodiaryxfr.php [hunk 2]
```diff
@@ -232,22 +232,18 @@
...
     $xfrresults = get_string('xfrresults', 'diary', ['one' => $xfrcountck , 'two' => $xfrcountxfrd]);
 } else {
     $xfrresults = '';
 }
-echo '<br><br><a href="'.$url1->out(false)
-    .'" class="btn btn-success" style="border-radius: 8px">'
-    .get_string('returnto', 'diary', $diary->name)
-    .'</a> '.$xfrresults;
+
+$link = html_writer::link($url1, get_string('returnto', 'diary', $diary->name), ['class' => 'btn btn-success', 'style' => 'border-radius: 8px']);
```
Comment (matthewhilton) [inline journaltodiaryxfr.php:247]: Does `use html_writer` need to be added to the top of this page ?

File: lib.php [hunk 3]
```diff
@@ -663,12 +663,12 @@ function diary_print_overview($courses, &$htmlarray) {
         }
 
         if ($diaryopen) {
-            $url = new moodle_url('mod/diary/view.php', ['id' => $diary->coursemodule]);
+            $url = new moodle_url('/mod/diary/view.php', ['id' => $diary->coursemodule]);
+            $link = html_writer::link($url, $diary->name, ['class' => $diary->visible ? '' : 'dimmed']);
```
Comment (matthewhilton) [inline lib.php:667]: does `use moodle_url` and `use html_writer` need to be added to the top of this page ?

File: prompt_edit.php [hunk 4]
```diff
@@ -182,7 +182,7 @@
         $data->minmaxparagraphpercent = $prompt->minmaxparagraphpercent;
 
         // If user can edit, create a delete link to the current prompt.
-        $url = new moodle_url('mod/diary/prompt_edit.php', ['id' => $id, 'action' => 'delete', 'promptid' => $prompt->id]);
+        $url = new moodle_url('/mod/diary/prompt_edit.php', ['id' => $id, 'action' => 'delete', 'promptid' => $prompt->id]);
```
Comment (matthewhilton) [inline prompt_edit.php:187]: does use moodle_url need to be added to the top of this page ?

File: report.php [hunk 5]
```diff
@@ -319,11 +319,8 @@
     $saveallbutton .= '<input type="submit" class="btn btn-primary" style="border-radius: 8px" value="'
                       .get_string("saveallfeedback", "diary").'" />';
     // 20200421 Added a return button.
-    $url = new moodle_url('mod/diary/view.php', ['id' => $id]);
-    $saveallbutton .= ' <a href="'.$url->out(false)
-                     .'" class="btn btn-secondary" role="button" style="border-radius: 8px">'
-                     .get_string('returnto', 'diary', $diary->name)
-                     .'</a>';
+    $url = new moodle_url('/mod/diary/view.php', ['id' => $id]);
+    $saveallbutton .= html_writer::link($url, get_string('returnto', 'diary', $diary->name), ['class' => 'btn btn-secondary', 'role' => 'button', 'style' => 'border-radius: 8px']);
```
Comment (matthewhilton) [inline report.php:325]: does use moodle_url and use html_writer need to be added to the top of this page ?

File: reportsingle.php [hunk 6]
```diff
@@ -174,7 +174,7 @@
 
     // 20211230 Tacked on an action for the return URL.
     // 20201222 Added a return to report.php button if you do not want to save feedback.
-    $url2 = new moodle_url('mod/diary/report.php', ['id' => $id, 'action' => 'currententry']);
+    $url2 = new moodle_url('/mod/diary/report.php', ['id' => $id, 'action' => 'currententry']);
```
Comment (matthewhilton) [inline reportsingle.php:178]: does use `moodle_url` need to be added to the top of this page ?

File: view.php [hunk 7]
```diff
@@ -236,7 +236,7 @@
 if (prompts::diary_available($diary)) {
     list($tcount, $past, $current, $future) = prompts::diary_count_prompts($diary);
     if ($current > 1) {
-        $url1 = new moodle_url('mod/diary/prompt_edit.php', ['id' => $cm->id]);
+        $url1 = new moodle_url('/mod/diary/prompt_edit.php', ['id' => $cm->id]);
```
Comment (matthewhilton) [inline view.php:240]: does use `moodle_url` need to be added to the top of this page ?

**Review by matthewhilton — CHANGES_REQUESTED** [review]
Looks good now - just a few small things :+1:  Feel free to ping me if you need any clarification

File: lib.php (hunk 3 above)
Comment (owenherbert) [reply lib.php:667]: Adding these to the top of the page seems to break the site:
```
Warning: The use statement with non-compound name 'moodle_url' has no effect in /var/www/evidence-central/mod/diary/lib.php on line 26

Warning: The use statement with non-compound name 'html_writer' has no effect in /var/www/evidence-central/mod/diary/lib.php on line 27
```

Comment (owenherbert) [reply lib.php:667]: Adding these to the top of the page seems to break the site:
```
Warning: The use statement with non-compound name 'moodle_url' has no effect in /var/www/evidence-central/mod/diary/lib.php on line 26

Warning: The use statement with non-compound name 'html_writer' has no effect in /var/www/evidence-central/mod/diary/lib.php on line 27
```

Comment (matthewhilton) [reply lib.php:667]: Ah yeah probably because this file isn't namespaced - don't worry about it then

**Pull request merged**